File: app/src/main/java/com/example/pam_app/repository/RoomBucketRepository.java
package com.example.pam_app.repository;

import com.example.pam_app.db.BucketDao;
import com.example.pam_app.db.BucketEntity;
import com.example.pam_app.db.BucketEntryWithBucketEntity;
import com.example.pam_app.model.Bucket;
import com.example.pam_app.model.BucketEntry;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Single;

import static com.example.pam_app.model.BucketType.SPENDING;

public class RoomBucketRepository implements BucketRepository {

    private final BucketDao bucketDao;
    private final BucketMapper bucketMapper;

    public RoomBucketRepository(BucketDao bucketDao, BucketMapper bucketMapper) {
        this.bucketDao = bucketDao;
        this.bucketMapper = bucketMapper;
    }

    @Override
    public Flowable<List<Bucket>> getList() {
        return this.bucketDao.getList().map(bucketEntityList -> {
            final List<Bucket> buckets = new ArrayList<>();
            for (final BucketEntity be: bucketEntityList) {
                buckets.add(bucketMapper.toModel(be));
            }
            return buckets;
        });
    }

    @Override
    public Flowable<List<Bucket>> getList(boolean type, Date date) {
        return this.bucketDao.getList(type, date).map(bucketEntityList -> {
            final List<Bucket> buckets = new ArrayList<>();
            for (final BucketEntity be: bucketEntityList) {
                buckets.add(bucketMapper.toModel(be));
            }
            return buckets;
        });
    }

    @Override
    public void update(Bucket bucket) {
        bucketDao.update(bucketMapper.toEntity(bucket));
    }

    @Override
    public Single<Long> create(Bucket bucket) {
        return Single.fromCallable(() -> bucketDao.create(bucketMapper.toEntity(bucket)));
    }

    @Override
    public Single<Integer> delete(int id) {
        return Single.fromCallable(() -> bucketDao.delete(id));
    }

    @Override
    public Flowable<Bucket> get(int id) {
        return this.bucketDao.get(id).map(bucketMapper::toModel);
    }

    @Override
    public Single<Bucket> get(final String title) {
        return this.bucketDao.getBucket(title).map(bucketMapper::toModel);
    }

    @Override
    public Flowable<List<BucketEntry>> getEntryList() {
        return this.bucketDao.getEntryList().map(bucketEntryEntityList -> {
            final List<BucketEntry> entries = new ArrayList<>();
            for (final BucketEntryWithBucketEntity be: bucketEntryEntityList) {
                entries.add(bucketMapper.toModel(be));
            }
            return entries;
        });
    }

    @Override
    public Flowable<List<Double>> getSpendingTotal() {
        return bucketDao.getTotalAmountByType(SPENDING.ordinal());
    }

    @Override
    public Single<Long> addEntry(BucketEntry entry, final int bucketId) {
        return Single.fromCallable(() -> bucketDao.addEntry(bucketMapper.toEntity(entry, bucketId)));
    }

    @Override
    public void removeEntry(BucketEntry entry, final int idBucket) {
        bucketDao.removeEntry(bucketMapper.toEntity(entry, idBucket));
    }

    @Override
    public Flowable<List<String>> getTitleListByType(int type) {
        return this.bucketDao.getTitleListByType(type);
    }
}

File: app/src/main/java/com/example/pam_app/repository/RoomBucketRepository.kt
package com.example.pam_app.repository

import android.database.sqlite.SQLiteConstraintException
import com.example.pam_app.db.BucketDao
import com.example.pam_app.db.BucketEntity
import com.example.pam_app.db.BucketEntryWithBucketEntity
import com.example.pam_app.db.BucketWithEntriesEntity
import com.example.pam_app.model.Bucket
import com.example.pam_app.model.BucketEntry
import com.example.pam_app.model.BucketType
import io.reactivex.Flowable
import io.reactivex.Single
import java.util.*
import kotlin.jvm.Synchronized

class RoomBucketRepository(private val bucketDao: BucketDao?, private val bucketMapper: BucketMapper?) : BucketRepository {
    override val list: Flowable<List<Bucket>>
        get() = bucketDao!!.list.map { bucketEntityList: List<BucketEntity?>? ->
            val buckets: MutableList<Bucket> = ArrayList()
            for (be in bucketEntityList!!) {
                buckets.add(bucketMapper!!.toModel(be))
            }
            buckets
        }

    override fun getList(type: Boolean, date: Date?): Flowable<List<Bucket?>> {
        return bucketDao!!.getList(type, date).map { bucketEntityList: List<BucketEntity?>? ->
            val buckets: MutableList<Bucket?> = ArrayList()
            for (be in bucketEntityList!!) {
                buckets.add(bucketMapper!!.toModel(be))
            }
            buckets
        }
    }

    override fun update(bucket: Bucket?): Single<Int?> {
        return Single.fromCallable { bucketDao!!.update(bucketMapper!!.toEntity(bucket))}
    }

    override fun create(bucket: Bucket?): Single<Long?> {
        return Single.fromCallable {
            try {
                return@fromCallable bucketDao!!.create(bucketMapper!!.toEntity(bucket))
            } catch (e: SQLiteConstraintException) {
                return@fromCallable -1L
            }
        }
    }

    override fun delete(id: Int): Single<Int?> {
        return Single.fromCallable { bucketDao!!.delete(id) }
    }

    override fun get(id: Int): Single<Bucket> {
        return bucketDao!![id].map { bucketEntity: BucketWithEntriesEntity? -> bucketMapper!!.toModel(bucketEntity) }
    }

    override fun get(title: String?): Single<Bucket> {
        return bucketDao!!.getBucket(title).map { bucketEntity: BucketEntity? -> bucketMapper!!.toModel(bucketEntity) }
    }

    override val entryList: Flowable<List<BucketEntry>>
        get() = bucketDao?.entryList?.map { bucketEntryEntityList: List<BucketEntryWithBucketEntity?>? ->
            val entries: MutableList<BucketEntry> = ArrayList()
            for (be in bucketEntryEntityList!!) {
                entries.add(bucketMapper!!.toModel(be))
            }
            entries
        }!!

    override val spendingTotal: Flowable<List<Double?>?>?
        get() = bucketDao!!.getTotalAmountByType(BucketType.SPENDING.ordinal)

    override fun addEntry(entry: BucketEntry, bucketId: Int): Single<Long?> {
        return Single.fromCallable { bucketDao!!.addEntry(bucketMapper!!.toEntity(entry, bucketId)) }
    }

    override fun removeEntry(entry: BucketEntry, idBucket: Int) {
        bucketDao!!.removeEntry(bucketMapper!!.toEntity(entry, idBucket))
    }

    override fun getTitleListByType(type: Int): Flowable<List<String?>?>? {
        return bucketDao!!.getTitleListByType(type)
    }

}